tests/test_conv_new.py

```diff
@@ -0,0 +1,323 @@
+    from distdl.nn.conv_feature_new import DistributedFeatureConv2d
+    from distdl.nn.conv_feature_new import DistributedFeatureConv3d
+    from distdl.nn.transpose import DistributedTranspose
+    from distdl.utilities.torch import zero_volume_tensor
+
+    # Isolate the minimum needed ranks
+    base_comm, active = comm_split_fixture
+    if not active:
+        return
+    P_world = MPIPartition(base_comm)
+
+    # Create the partitions
+    P_root_base = P_world.create_partition_inclusive(np.arange(1))
+    P_x_base = P_world.create_partition_inclusive(P_x_ranks)
+    P_root = P_root_base.create_cartesian_topology_partition([1]*len(P_x_shape))
+    P_x = P_x_base.create_cartesian_topology_partition(P_x_shape)
+
+    # Create the layers
+    if input_dimensions == 1:
+        dist_layer_type = DistributedFeatureConv1d
+        seq_layer_type = Conv1d
+    elif input_dimensions == 2:
+        dist_layer_type = DistributedFeatureConv2d
+        seq_layer_type = Conv2d
+    elif input_dimensions == 3:
+        dist_layer_type = DistributedFeatureConv3d
+        seq_layer_type = Conv3d
+    scatter = DistributedTranspose(P_root, P_x)
+    dist_layer = dist_layer_type(P_x,
+                                 in_channels=x_global_shape[1],
+                                 out_channels=10,
+                                 kernel_size=kernel_size,
+                                 padding=padding,
+                                 stride=stride,
+                                 dilation=dilation,
+                                 bias=bias)
+    gather = DistributedTranspose(P_x, P_root)
+    if P_root.active:
+        seq_layer = seq_layer_type(in_channels=x_global_shape[1],
+                                   out_channels=10,
+                                   kernel_size=kernel_size,
+                                   padding=padding,
+                                   stride=stride,
+                                   dilation=dilation,
+                                   bias=bias)
+        # set the weights of both layers to be the same
+        weight = torch.rand_like(seq_layer.weight)
+        seq_layer.weight.data = weight
+        dist_layer.weight.data = weight
+        if bias:
+            bias_weight = torch.rand_like(seq_layer.bias)
+            seq_layer.bias.data = bias_weight
+            dist_layer.bias.data = bias_weight
+
+    # Create the input
+    if P_root.active:
+        x = np.random.randn(*x_global_shape)
+        dist_x = torch.from_numpy(x.copy()).to(torch.float32)
+        seq_x = torch.from_numpy(x.copy()).to(torch.float32)
+        dist_x.requires_grad = True
+        seq_x.requires_grad = True
+    else:
+        dist_x = zero_volume_tensor(requires_grad=True)
+
+    # Check the forward pass
+    dist_y = gather(dist_layer(scatter(dist_x)))
+    if P_root.active:
+        seq_y = seq_layer(seq_x)
+        assert dist_y.shape == seq_y.shape
+        assert torch.allclose(dist_y, seq_y)
```

Comment:
See the upsampling test (https://github.com/distdl/distdl/blob/master/tests/test_upsampling.py#L375) for the (safe) way to test allclose with torch tensors.

Reply:
Just for future reference, if this is threshold is not set properly, then certain implementations of MPI that are not compiled in a float-safe manner will likely fail at least some of the time. This was something I ran into when testing.

Reply:
What do you mean?

Reply:
I was just elaborating on why setting atol forgivingly (~sqrt(e_machine)) is better. Otherwise, tests may fail on some MPI implementations compiled with -Ofast or other optimizations that do not preserve floating point accuracy. And since inputs for tests are randomized without a seed, they may fail/pass in an undefined manner.